### ISSUE TITLE
水印图像视觉质量

### ISSUE THREAD

**Issue by Uranium-Deng** [jzyustc/MBRS#6]
你好，读了你的论文，受益匪浅，非常感谢！
我用你本仓库的代码训练模型时遇到一些疑惑，想咨询一下作者

1. **水印图像能看到明显的改动痕迹**
(1) 论文中生成的水印图像的PSNR和SSIM值都比较高，我在重新训练模型时也验证了这一点(使用学习率衰减训练策略后的PSNR值比论文中还高)，虽然PSNR和SSIM值都比较高，但是观察生成的水印图像，还是能明显的看出改动痕迹，有明显的亮点，如下图。
![img1](https://user-images.githubusercontent.com/56220676/220548526-1cce2c80-a1b3-4992-b85f-47ab18725163.jpg)
![img2](https://user-images.githubusercontent.com/56220676/220548667-5f711cb8-072c-49db-8e28-dc8b94bd5807.jpg)

(2) 为保证模型泛化和迁移性，作者在图像输入网络前做了数据增强，to_tensor()归一化后，采用Normalize([0.5, 0.5, 0.5], [0.5, 0.5, 0.5])进行正则化，此处我修改为Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]，不知为何，用后者Normalize，生成的水印图像除了包含不规则的改动，还会包含<font color=red>**规则的网格状的亮点（如下图）**</font>，而前者Normalize只有一些不规则的改动。想问作者是否遇到这个问题。
![img1](https://user-images.githubusercontent.com/56220676/220548892-33b42a2d-40e5-4dee-bce3-9054eadb86a5.jpg)
![Lena](https://user-images.githubusercontent.com/56220676/220549106-3c662f93-eca6-4eda-986d-7636fe4f2464.png)

(3) 代码中计算PSNR和SSIM，是在模型输出的图像和数据增强后的图像之间计算的，这样说不上错，但是个人认为直接计算 原始图像和水印图像的PSNR和SSIM更好，毕竟模型输出的图张量 还需要 reverse normalize 才能转换为 水印图像。

2. **误码率 error_rate**
(1) 论文中的误码率，由代码判断采用的似乎是平均误码率，即将计算每一个message的误码率，然后加和求平均值。个人认为若换一种计算方式：用提取水印错误的图像的数量 除以 总水印图像的数量，这样的计算方式更具有说明性。想问一下作者，水印论文中计算误码率采用的方式都是前者吗？

感谢！:)

**Comment by jzyustc** [comment]
你好，感谢你对我们工作的关注！

1. 1）我们在训练中没有发现过图片中这么明显的修改痕迹。请问使用repo提供的pretrain模型能否复现这个问题？如果能，请麻烦提供一下原图，我测试一下问题所在；如果不能，那这可能是修改代码/重新训练时出现的问题，可以再check一下代码。

1. 2）代码中 Noise Layer 等部分存在很多和 image 取值范围相关的操作。例如在JPEG模拟函数中：https://github.com/jzyustc/MBRS/blob/c0f31767bfe4262ea633c5779dacda6264795f7a/network/noise_layers/jpeg.py#L165
rgb到yuv的色域转变就是从[-1, 1]转移的。如果需要修改normalization，需要对这些地方都进行统一修改。

1. 3） 在测试时确实不应该进行resize, crop等数据增强，很抱歉造成了误导。Normalization 对实验结果仅有浮点误差量级的影响，是否进行 reverse normalize 对实验结果几乎没有影响。

3. 1）paper 里采用的是 bit error rate 为指标，因此计算了每个 bit 的平均正确率。我见到的大多数论文都是以 bit accuracy 衡量的，当然你也可以使用 message accuracy 衡量，只要统一标准就可以。

**Comment by Uranium-Deng** [comment]
你好，感谢回复！
针对水印图像的视觉质量，我用本repo提供的pretain模型（MBRS_256_m256）做了实验，实现结果显示：部分图像视觉效果是OK的，但也有部分图像仍存在明显亮点。实验结果图片如下(图片来自mirflickr25k，和StegaStamp是同一个数据集)：
![2023-02-23 15-18-32 的屏幕截图](https://user-images.githubusercontent.com/56220676/220843187-17811eb9-613b-435c-881e-70e635bb010b.png)
![2023-02-23 15-19-04 的屏幕截图](https://user-images.githubusercontent.com/56220676/220843221-f9d8867b-9cba-4402-9145-78f653ed4d20.png)
![2023-02-23 15-19-48 的屏幕截图](https://user-images.githubusercontent.com/56220676/220843247-4755eb2e-51b2-48fd-865d-77529ef705b2.png)
![2023-02-23 15-20-14 的屏幕截图](https://user-images.githubusercontent.com/56220676/220843261-94af4010-e186-44e8-b1fd-9c8f74b5d80b.png)
![2023-02-23 15-21-01 的屏幕截图](https://user-images.githubusercontent.com/56220676/220843272-fd96eefb-4234-4d7c-b6dc-15a0b4f59d93.png)

这是我的部分代码：
~~~python
def load_model():
    network = Network(H=height, W=width, message_length=message_length, noise_layers=["JpegTest(50)"],
                      device=device, batch_size=1, lr=1e-3, with_diffusion=False,
                      only_decoder=False, scheduler_epoch=20, scheduler_gamma=0.2)
    EC_load_path = "/home/ruizhi/work/Git/model_MBRS/results/MBRS_256_m256/models/EC_42.pth"
    Dis_load_path = "/home/ruizhi/work/Git/model_MBRS/results/MBRS_256_m256/models/D_42.pth"
    network.source_load_model(EC_load_path, Dis_load_path)
    print("load model successfully")
    return network


def encode():
    # 待嵌入图像 和 待嵌入信息
    strength_factor = 1
    cover_image_path = "/home/ruizhi/work/Git/model_MBRS/images/cover_image/Lena.png"  # 图片尺寸需匹配网络输入尺寸
    image_name = cover_image_path.split('/')[-1]
    source_cover_image = np.asarray(Image.open(cover_image_path).convert("RGB").resize((height, width)))
    transform_image = transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize(mean=[0.5, 0.5, 0.5], std=[0.5, 0.5, 0.5])
    ])
    cover_image = transform_image(source_cover_image)
    cover_image = cover_image.unsqueeze(0)
    message = torch.Tensor(np.random.choice([0, 1], (cover_image.shape[0], message_length))).to(device)
    print("embed message: \n", message)

    # 模型加载
    network = load_model()

    network.encoder_decoder.eval()
    network.discriminator.eval()
    with torch.no_grad():
        # use device to compute
        cover_image, messages = cover_image.to(network.device), message.to(network.device)
        encoded_image = network.encoder_decoder.module.encoder(cover_image, messages)
        encoded_image = cover_image + (encoded_image - cover_image) * strength_factor

        psnr = -kornia.losses.psnr_loss(encoded_image.detach(), cover_image, 2).item()
        ssim = 1 - 2 * kornia.losses.ssim_loss(encoded_image.detach(), cover_image, window_size=5,
                                               reduction="mean").item()

        plt.subplot(1, 3, 1)
        plt.title("cover_image")
        plt.imshow(source_cover_image)
        plt.subplot(1, 3, 2)
        plt.title("encoded_image, PSNR: %.3f, SSIM: %.3f" % (psnr, ssim))
        encoded_image = encoded_image.detach().cpu().squeeze(0).permute(1, 2, 0).numpy()

        # scale values to range [0, 1] from original range of [-1, 1]
        # encoded_image = (encoded_image + 1) / 2

        # 反Normalize操作, 网络训练的groundTruth是标准化的图像
        encoded_image = (encoded_image * [0.5, 0.5, 0.5] + [0.5, 0.5, 0.5]) * 255

        encoded_image = encoded_image.astype(np.uint8)
        encoded_image = np.clip(encoded_image, 0, 255)
        plt.imshow(encoded_image)

        residual_img = np.abs(encoded_image - source_cover_image)
        plt.subplot(1, 3, 3)
        plt.title("residual_image")
        plt.imshow(residual_img)

        plt.show()
        print("encoded_image, PSNR: %.3f, SSIM: %.3f" % (psnr, ssim))
        # save encoded image
        im = Image.fromarray(encoded_image) 
        im.save(os.path.join("/home/ruizhi/work/Git/model_MBRS/images", image_name), quality=100)
~~~

此外仓库代码utils/save_images.py的save_images()函数中有一段代码如下：
~~~python
# scale values to range [0, 1] from original range of [-1, 1]
# encoded_image = (encoded_image + 1) / 2
~~~
该行代码将图像值域从 [-1, 1] 变为 [0, 1]，方便后续处理。但是个人认为，这里的处理更换为：
~~~python
encoded_image = (encoded_image * [0.5, 0.5, 0.5] + [0.5, 0.5, 0.5]) (超出[0, 1]范围的，可直接torch.clamp截断)
~~~
或许会更好，因为Encoder输出图像的ground trouth就是cover image经过normalize之后的张量。
感谢！:)

**Comment by jzyustc** [comment]
@Uranium-Deng 你这些demo里的结果肯定是有问题的，我在实验里没有看到这种明显的伪影。
我看了一下，估计问题应该在这两行代码：
```
encoded_image = encoded_image.astype(np.uint8)
encoded_image = np.clip(encoded_image, 0, 255)
```
正确的做法应该是先clip再转为uint8：
```
encoded_image = np.clip(encoded_image, 0, 255)
encoded_image = encoded_image.astype(np.uint8)
```
否则原本大于255 / 小于0的值会被错误的转化。具体可以去google一下uint8格式。

**Comment by Uranium-Deng** [comment]
老天，确实是这两行代码出现了问题！
![2023-02-23 16-18-29 的屏幕截图](https://user-images.githubusercontent.com/56220676/220854514-161e511b-d43c-4818-b9b3-df72c2415bbd.png)
太感谢了，比心比心 :)))))))))))))

**Comment by Uranium-Deng** [comment]
确实是这两行代码的问题，我尝试了自己训练的模型，图像的视觉质量也挺好的。
![2023-02-23 16-26-35 的屏幕截图](https://user-images.githubusercontent.com/56220676/220856271-3719cb44-ce04-4968-b195-4e0668d9c17b.png)
![2023-02-23 16-27-28 的屏幕截图](https://user-images.githubusercontent.com/56220676/220856289-f1a3ae88-8bc9-4160-bd62-7d6993dd920b.png)
感谢作者！我一直在Debug调参，还以为是我训练的方式有问题，没想到问题出现在这个小地方，23333。

**Comment by jzyustc** [comment]
Issue solved. Close this comment.
